internal/form/submit/service.go

```diff
@@ -0,0 +1,100 @@
+package submit
+
+import (
+	"NYCU-SDC/core-system-backend/internal/form/question"
+	"NYCU-SDC/core-system-backend/internal/form/response"
+	"NYCU-SDC/core-system-backend/internal/form/shared"
+	"context"
+	"fmt"
+	logutil "github.com/NYCU-SDC/summer/pkg/log"
+	"github.com/google/uuid"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/trace"
+	"go.uber.org/zap"
+)
+
+type QuestionStore interface {
+	ListByFormID(ctx context.Context, formID uuid.UUID) ([]question.Answerable, error)
+}
+
+type FormResponseStore interface {
+	CreateOrUpdate(ctx context.Context, formID uuid.UUID, userID uuid.UUID, answers []shared.AnswerParam, questionType []response.QuestionType) (response.FormResponse, error)
+}
+
+type Service struct {
+	logger *zap.Logger
+	tracer trace.Tracer
+
+	questionStore QuestionStore
+	responseStore FormResponseStore
+}
+
+func NewService(logger *zap.Logger, questionStore QuestionStore, formResponseStore FormResponseStore) *Service {
+	return &Service{
+		logger:        logger,
+		tracer:        otel.Tracer("submit/service"),
+		questionStore: questionStore,
+		responseStore: formResponseStore,
+	}
+}
+
+// Submit handles a user's submission for a specific form.
+// It performs the following steps:
+// 1. Retrieves all questions associated with the form.
+// 2. Validates the submitted answers against the corresponding questions.
+//   - If any validation fails or if an answer references a nonexistent question, it accumulates the errors.
+//
+// 3. If there are validation errors, returns them without saving.
+// 4. If validation passes, creates or updates the response record using the answer values and question types.
+//
+// Returns the saved form response if successful, or a list of validation/database errors otherwise.
+func (s *Service) Submit(ctx context.Context, formID uuid.UUID, userID uuid.UUID, answers []shared.AnswerParam) (response.FormResponse, []error) {
+	traceCtx, span := s.tracer.Start(ctx, "Submit")
+	defer span.End()
+	logger := logutil.WithContext(traceCtx, s.logger)
+
+	questions, err := s.questionStore.ListByFormID(traceCtx, formID)
+	if err != nil {
+		return response.FormResponse{}, []error{err}
+	}
+
+	// Validate answers against questions
+	questionTypes := make([]response.QuestionType, len(questions))
```

Comment:
This would create a slice initialized with zero values with length equal to the total number of questions, leading to the number of question types would be twice of the number of answer type after the append function. Thus, consider createing an empty slice with capacity for the number of answers to solve the mismatch issue.
```suggestion
	questionTypes := make([]response.QuestionType, 0, len(questions))
```